apstools/devices/area_detector_factory.py

```diff
@@ -0,0 +1,330 @@
+        "suffix": "ROI3:",
+        "class": ophyd.areadetector.plugins.ROIPlugin_V34,
+    },
+    "roi4": {
+        "suffix": "ROI4:",
+        "class": ophyd.areadetector.plugins.ROIPlugin_V34,
+    },
+    "roistat1": {
+        "suffix": "ROIStat1:",
+        "class": ophyd.areadetector.plugins.ROIStatPlugin_V34,
+    },
+    "scatter1": {
+        "suffix": "Scatter1:",
+        "class": ophyd.areadetector.plugins.ScatterPlugin_V34,
+    },
+    "stats1": {
+        "suffix": "Stats1:",
+        "class": ophyd.areadetector.plugins.StatsPlugin_V34,
+    },
+    "stats2": {
+        "suffix": "Stats2:",
+        "class": ophyd.areadetector.plugins.StatsPlugin_V34,
+    },
+    "stats3": {
+        "suffix": "Stats3:",
+        "class": ophyd.areadetector.plugins.StatsPlugin_V34,
+    },
+    "stats4": {
+        "suffix": "Stats4:",
+        "class": ophyd.areadetector.plugins.StatsPlugin_V34,
+    },
+    "stats5": {
+        "suffix": "Stats5:",
+        "class": ophyd.areadetector.plugins.StatsPlugin_V34,
+    },
+    "transform1": {
+        "suffix": "Trans1:",
+        "class": ophyd.areadetector.plugins.TransformPlugin_V34,
+    },
+    # file writers
+    "hdf1": {
+        "class": HDF5FileWriterPlugin,
+        "read_path_template": None,
+        "suffix": "HDF1:",
+        "write_path_template": None,
+    },
+    "jpeg1": {
+        "class": AD_EpicsFileNameJPEGPlugin,
+        "read_path_template": None,
+        "suffix": "JPEG1:",
+        "write_path_template": None,
+    },
+    "magick1": {
+        "class": ophyd.areadetector.plugins.MagickPlugin_V34,
+        "read_path_template": None,
+        "suffix": "Magick1:",
+        "write_path_template": None,
+    },
+    "netcdf1": {
+        "class": ophyd.areadetector.plugins.NetCDFPlugin_V34,
+        "read_path_template": None,
+        "suffix": "netCDF1:",
+        "write_path_template": None,
+    },
+    "tiff1": {
+        "class": AD_EpicsFileNameTIFFPlugin,
+        "read_path_template": None,
+        "suffix": "TIFF1:",
+        "write_path_template": None,
+    },
+}
+"""Default plugin configuration dictionary."""
+
+
+def ad_class_factory(name, bases=None, plugins=None, plugin_defaults=None):
+    """
+    Build an Area Detector class with specified plugins.
+
+    PARAMETERS
+
+    name
+        *str* :
+        Name of the class to be created.
+    bases
+        *object* or *tuple* :
+        Parent(s) of the new class.
+        (default: ``(SingleTrigger_V34, DetectorBase)``)
+    plugins
+        *list* :
+        Description of the plugins used.
+        (default: ``["cam"]`` -- Just the camera plugin.)
+    plugin_defaults
+        *object*:
+        Plugin configuration dictionary.
+        (default: ``None``, PLUGIN_DEFAULTS will be used.)
+    """
+    bases = bases or (
+        SingleTrigger_V34,
+        ophyd.areadetector.DetectorBase,
+    )
+    plugins = plugins or ["cam"]
+    plugin_defaults = plugin_defaults or PLUGIN_DEFAULTS
```

Comment:
This makes it impossible to pass an empty dict (``{}``) to this function as *plugin_defaults*, since ``bool({})`` evaluates to ``False`` and then ``PLUGIN_DEFAULTS`` would get used.

Maybe not a big deal since I can't imagine many people would want this. But I think it could be avoided by using ``PLUGIN_DEFAULTS`` as the default value for the *plugin_defaults*, which makes the intention more clear when looking at the call signature.

Reply:
Can do.

Reply:
Same for `plugins=None`.  Changing it to `plugins=["cam"]`.

Reply:
Ah... there's the reason to keep the default kwarg value of `None` -- in the unit tests, `None` is used.

```py
ERROR apstools/devices/tests/test_area_detector_support.py::test_AD_EpicsFileNameMixin[hdf1-AD_HDF5] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_AD_EpicsFileNameMixin[jpeg1-AD_JPEG] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_AD_EpicsFileNameMixin[tiff1-AD_TIFF] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_stage[hdf1-AD_EpicsFileNameHDF5Plugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_stage[jpeg1-AD_EpicsFileNameJPEGPlugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_stage[tiff1-AD_EpicsFileNameTIFFPlugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_bp_count_custom_name[hdf1-AD_EpicsFileNameHDF5Plugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_bp_count_custom_name[jpeg1-AD_EpicsFileNameJPEGPlugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_bp_count_custom_name[tiff1-AD_EpicsFileNameTIFFPlugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_full_file_name_local - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_no_file_path - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_file_numbering - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_capture_error_with_single_mode - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_single_mode - TypeError: type.__new__() argument 2 must be tuple, not None
```

Code like this is needed:

```py
    if plugin_defaults is None:
        plugin_defaults = PLUGIN_DEFAULTS
```

Reply:
Okay, sounds good.

Reply:
> I can't imagine many people would want `plugin_defaults={}`

Actually, this is a realistic case to force the caller to define all keywords of each plugin.  Probably only useful for unit testing, not real world?

Reply:
@canismarko Will these changes work for you?

Reply:
Yes, changes look good to me. Thanks!

Reply:
I also added some documentation (look for `EXAMPLE 1:` and others).